OvmfPkg/PlatformPei/Platform.c

```diff
@@ -198,6 +199,29 @@ MemMapInitialization (
       ASSERT (TopOfLowRam <= PciExBarBase);
       ASSERT (PciExBarBase <= MAX_UINT32 - SIZE_256MB);
       PciBase = (UINT32)(PciExBarBase + SIZE_256MB);
+    } else if (mHostBridgeDevId == VIRT_QEMU_DEVICE_ID) {
+      FIRMWARE_CONFIG_ITEM  FwCfgItemPciExBar;
+      UINTN                 FwCfgSizePciExBar;
+      FIRMWARE_CONFIG_ITEM  FwCfgItemSegment;
+      UINTN                 FwCfgSizeSegment;
+      UINT64                PciInfo[2];
+      UINT16                Segment;
+
+      QemuFwCfgFindFile ("etc/pci-host/base", &FwCfgItemPciExBar, &FwCfgSizePciExBar);
+      QemuFwCfgSelectItem (FwCfgItemPciExBar);
+      QemuFwCfgReadBytes (sizeof(PciInfo), &PciInfo);
+      PciExBarBase = PciInfo[0];
+      PciBase = PciInfo[1];
+      ASSERT (TopOfLowRam <= PciExBarBase);
+      ASSERT (PciExBarBase <= MAX_UINT32 - SIZE_256MB);
+      ASSERT (PciBase <= MAX_UINT32 - SIZE_256MB);
+
+      QemuFwCfgFindFile ("etc/pci-host/total-segment", &FwCfgItemSegment, &FwCfgSizeSegment);
+      QemuFwCfgSelectItem (FwCfgItemSegment);
+      QemuFwCfgReadBytes (sizeof(Segment), &Segment);
+      ASSERT (Segment <= MAX_UINT16);
+
+      PciExBarSize = Segment * SIZE_1MB;

```

Comment:
@liujing2 we should not make this assumption. The number of buses per segment should be configurable. MCFG allows each segment to have variable number of buses. We should keep this code generic and reflect what the PCI spec actually allows.

Reply:
@mcastelino Thanks for the comment. So let's pass the bus number of each segment via fw_cfg? For now, I didn't arrange file for every segment because it seems we only need to know the first base of mcfg and hole, ovmf calculates the space for multiple segment together.

Reply:
@liujing2 this means you assume that all the segments are contiguous. Is it possible for the segments to be non contiguous? If we supported that, the memory layout become more flexible, including placing some segments higher than 32bit.